### PR TITLE
Treat afterLastDeadline:null as "No access", differentiate from Closed (no submissions)

### PR DESCRIPTION
**Pull request by austinbillings-pl**
# Description

Previously, omitting `afterLastDeadline` and setting `allowSubmissions: false` were treated identically (authorized but not submittable). Now omitting `afterLastDeadline` (`null`) means the student loses access entirely after the last deadline, while `allowSubmissions=false` remains a view-only/review state.

- Add "No access" as the first (default) option in the AfterLastDeadline dropdown, mapping to afterLastDeadline: null
- Add `accessible` field to AccessTimelineEntry; set to false when afterLastDeadline is omitted/null
- Deny access in the resolver when the current segment is not accessible
- Accept nullable afterLastDeadline in the JSON schema (for overrides)
- Update RuleSummary to display "No access" vs "Closed" distinctly
- Add `date_control_after_last_deadline_overridden` column so overrides can distinguish "not overridden" (inherit default) from "overridden to no access" (`null`). Without this, both cases stored `allow_submissions=NULL` in the DB, so the default rule's value always bled through during merge.
- Update sync, tRPC, and model layers to read/write the new overridden flag
- Add model tests for the overridden flag round-trip (null override, not-overridden, legacy compat)
- [X] I have disclosed the level of AI assistance used: majority

# Testing
- tested setting, sync, and UI for all afterLastDeadline options in UI, observed correct behavior on student side
- all tests pass

### CODE REVIEW

**Comment by nwalters512** [comment]
What's the story on overrides here? That is, how does one differentiate between "not overridden" and "overridden to no access"?

**Comment by austinbillings-pl** [comment]
> What's the story on overrides here? That is, how does one differentiate between "not overridden" and "overridden to no access"?

@nwalters512 you've def pointed out an issue here; it seems that the fidelity of overriding-afterLastDeadline is lost as for example `{ dueDate: '...', afterLastDeadline: { allowSubmissions: true }` with an override `{ afterLastDeadline: null }` will get smashed when child props are checked, a la sync line 93:
`const afterLastDeadlineAllowSubmissionsField = mapField(afterLastDeadline?.allowSubmissions);`

 Claude posits that we could add a `date_control_after_last_deadline_overridden` db col to track overrides, with this plan:
   1. DB migration: Add `date_control_after_last_deadline_overridden` column
  2. Sync: Use `mapField(afterLastDeadline)` instead of `mapField(afterLastDeadline?.allowSubmissions)` to detect override
  3. data.ts: Use the new overridden flag to set `afterLastDeadline: null` when overridden but allow_submissions is null
  4. tRPC access-control.ts: Same
  5. DB types: Update Zod schema and table description

**Comment by nwalters512** [comment]
IMO the `_overridden` column is probably what we have to do here. Can you raise this for quicker discussion in the Slack channel to make sure this is the best path forward?

**Comment by reteps** [comment]
Given that we need a DB column, this should still be in draft.

File: apps/prairielearn/src/models/assessment-access-control-rules.ts [hunk 1]
```diff
@@ -289,6 +303,7 @@ export async function syncEnrollmentAccessControl(
     date_control_due_credit: ruleData.dueCredit,
     date_control_early_deadlines_overridden: ruleData.earlyDeadlinesOverridden,
     date_control_late_deadlines_overridden: ruleData.lateDeadlinesOverridden,
+    date_control_after_last_deadline_overridden: ruleData.afterLastDeadlineOverridden,
     date_control_after_last_deadline_allow_submissions: ruleData.afterLastDeadlineAllowSubmissions,
     date_control_after_last_deadline_credit: ruleData.afterLastDeadlineCredit,
```
[the remark this answers is not in the document]
Comment (austinbillings-pl) [reply apps/prairielearn/src/models/assessment-access-control-rules.ts]: addressed in 66b34e765

File: apps/prairielearn/src/lib/assessment-access-control/data.ts [hunk 2]
```diff
@@ -89,7 +89,17 @@ function buildDateControl(
...
+  if (rule.date_control_after_last_deadline_overridden) {
+    if (rule.date_control_after_last_deadline_allow_submissions != null) {
+      dateControl.afterLastDeadline = {
+        allowSubmissions: rule.date_control_after_last_deadline_allow_submissions,
+        credit: rule.date_control_after_last_deadline_credit,
+      };
+    } else {
+      dateControl.afterLastDeadline = null;
+    }
+  } else if (rule.date_control_after_last_deadline_allow_submissions != null) {
```
Comment (EduardoMVAz) [inline apps/prairielearn/src/lib/assessment-access-control/data.ts:101]: Claude left this comment here:

```
The legacy fallback in `dbBaseRowToAccessControlJson` is asymmetric with the one in `data.ts`, which I think causes a silent behavior flip for pre-migration rows.

- `apps/prairielearn/src/lib/assessment-access-control/data.ts` — legacy path checks `allow_submissions != null`, so `false` round-trips as `{ allowSubmissions: false }` (view-only).
- `apps/prairielearn/src/models/assessment-access-control-rules.ts` — legacy path checks `allowSubmissions === true`. When the column is `false`, it falls through and emits no `afterLastDeadline` at all, which under the new semantics means "no access."

Net effect: a course instance synced before this PR with `afterLastDeadline: { allowSubmissions: false }` will be read as **view-only by the resolver** but **rendered/exported as "no access" by the form and JSON layer**. The two views of the same row disagree.

Two options:

1. Mirror `data.ts` and handle `allowSubmissions === false` in the legacy branch of `dbBaseRowToAccessControlJson` so both readers agree.
2. Backfill `date_control_after_last_deadline_overridden = true` for existing rows where `allow_submissions IS NOT NULL` in the same migration, then drop the legacy branches entirely.

I'd lean toward (2) — two divergent legacy code paths is the kind of thing that's going to bite someone later. Either way, a unit test covering the `allow_submissions=false, overridden=false` row in `dbBaseRowToAccessControlJson` would lock this down.
```

Verify if it should be addressed

File: apps/prairielearn/src/pages/instructorAssessmentAccess/components/RuleSummary.tsx [hunk 3]
```diff
@@ -528,9 +532,9 @@ function formatAfterLastDeadline(afterLastDeadline: AfterLastDeadlineValue): str
     parts.push(`${afterLastDeadline.credit}% credit`);
   }
   if (afterLastDeadline.allowSubmissions) {
-    parts.push('submissions allowed');
+    parts.push(parts.length > 0 ? 'submissions allowed' : 'Submissions allowed');
```
Comment (EduardoMVAz) [inline apps/prairielearn/src/pages/instructorAssessmentAccess/components/RuleSummary.tsx:495]: Why does this have a different case for each of the possiblities?

File: apps/prairielearn/src/schemas/schemas/infoAssessment.json [hunk 4]
```diff
@@ -824,19 +824,26 @@
...
             },
             "afterLastDeadline": {
-              "description": "Controls for assessment behavior after last deadline",
-              "type": "object",
-              "additionalProperties": false,
-              "required": ["allowSubmissions"],
-              "properties": {
-                "allowSubmissions": {
-                  "type": "boolean"
+              "description": "Controls for assessment behavior after last deadline. Null means no access; omitted means inherit from default (overrides only).",
```
Comment (EduardoMVAz) [inline apps/prairielearn/src/schemas/schemas/infoAssessment.json]: Shouldn't you mention that omitted in the default rule means view only?

File: apps/prairielearn/src/schemas/accessControl.ts [hunk 5]
```diff
@@ -57,9 +57,11 @@ const DateControlJsonSchema = z
       .nullable()
       .optional()
       .describe('Array of late deadlines with credit as percentages'),
-    afterLastDeadline: AfterLastDeadlineJsonSchema.describe(
-      'Controls for assessment behavior after last deadline',
-    ).optional(),
+    afterLastDeadline: AfterLastDeadlineJsonSchema.nullable()
+      .describe(
+        'Controls for assessment behavior after last deadline. Null means no access; omitted means inherit from default (overrides only).',
```
Comment (EduardoMVAz) [inline apps/prairielearn/src/schemas/accessControl.ts]: Shouldn't you mention that omitted in the default rule means view only?

File: apps/prairielearn/src/pages/instructorAssessmentAccess/components/RuleSummary.tsx (hunk 3 above)
Comment (EduardoMVAz) [reply apps/prairielearn/src/pages/instructorAssessmentAccess/components/RuleSummary.tsx:495]: Oh, I saw the rendered value.

**Review by EduardoMVAz — APPROVED** [review]
LGTM, I would address this Claude comment, but I was not sure that it was right, so approving and leaving it up to you to verify.

File: apps/prairielearn/src/lib/assessment-access-control/data.ts (hunk 2 above)
Comment (austinbillings-pl) [reply apps/prairielearn/src/lib/assessment-access-control/data.ts:101]: I think we're good here; the only difference seems to be emitting `credit: null` or not, which is inconsequential since it's view-only regardless 👌

File: apps/prairielearn/src/schemas/schemas/infoAssessment.json (hunk 4 above)
Comment (austinbillings-pl) [reply apps/prairielearn/src/schemas/schemas/infoAssessment.json]: good call, resolved in f2c386cf2

File: apps/prairielearn/src/schemas/accessControl.ts (hunk 5 above)
Comment (austinbillings-pl) [reply apps/prairielearn/src/schemas/accessControl.ts]: as mentioned, improved in f2c386cf2
